theano/tensor/signal/roi_pool.c

```diff
@@ -0,0 +1,240 @@
+        int c = (index / pooled_width / pooled_height) % channels;
+        int n = index / pooled_width / pooled_height / channels;
+        bottom_rois += n * 5;
+        int roi_batch_index = bottom_rois[0];
+        int roi_start_w = round(bottom_rois[1] * spatial_scale);
+        int roi_start_h = round(bottom_rois[2] * spatial_scale);
+        int roi_end_w = round(bottom_rois[3] * spatial_scale);
+        int roi_end_h = round(bottom_rois[4] * spatial_scale);
+        int roi_width = max(roi_end_w - roi_start_w + 1, 1);
+        int roi_height = max(roi_end_h - roi_start_h + 1, 1);
+        float bin_size_h = (float)(roi_height)
+                           / (float)(pooled_height);
+        float bin_size_w = (float)(roi_width)
+                           / (float)(pooled_width);
+
+        int hstart = (int)(floor((float)ph * bin_size_h));
+        int wstart = (int)(floor((float)pw * bin_size_w));
+        int hend = (int)(ceil((float)(ph + 1)
+                                         * bin_size_h));
+        int wend = (int)(ceil((float)(pw + 1)
+                                         * bin_size_w));
+
+        // Add roi offsets and clip to input boundaries
+        hstart = min(max(hstart + roi_start_h, 0), height);
+        hend = min(max(hend + roi_start_h, 0), height);
+        wstart = min(max(wstart + roi_start_w, 0), width);
+        wend = min(max(wend + roi_start_w, 0), width);
+        bool is_empty = (hend <= hstart) || (wend <= wstart);
+
+        // Define an empty pooling region to be zero
+        float maxval = is_empty ? 0 : -1000;
+        // If nothing is pooled, argmax = -1 causes nothing to be backprop'd
+        int maxidx = -1;
+        bottom_data += (roi_batch_index * channels + c) * height * width;
+        for (int h = hstart; h < hend; ++h) {
+            for (int w = wstart; w < wend; ++w) {
+            int bottom_index = h * width + w;
+                if (bottom_data[bottom_index] > maxval) {
+                  maxval = bottom_data[bottom_index];
+                  maxidx = bottom_index;
+                }
+            }
+        }
+        top_data[index] = maxval;
+        argmax_data[index] = maxidx;
+    }
+}
+
+int APPLY_SPECIFIC(CPUFwd)(PyArrayObject* data,
+                      PyArrayObject* rois,
+                      PyArrayObject** out,
+                      PyArrayObject** argmaxes) {
+  int batch_size = PyArray_DIMS(rois)[0];
+  int channels = PyArray_DIMS(data)[1];
+  int height = PyArray_DIMS(data)[2];
+  int width = PyArray_DIMS(data)[3];
+  int data_typenum = PyArray_ObjectType((PyObject*)(data), 0);
+  // Prepare outputs.
+  int dims[] = {0, 0, 0, 0};
+  dims[0] = batch_size;
+  dims[1] = channels;
+  dims[2] = POOLED_HEIGHT;
+  dims[3] = POOLED_WIDTH;
+  int total_ndim = 4;
+  int count = batch_size * channels * POOLED_HEIGHT * POOLED_WIDTH;
+  int mem_nc;
+  mem_nc = 0;
+  // Checking if contiguous
+  if(!PyArray_ISCONTIGUOUS(data) || !PyArray_ISCONTIGUOUS(rois)){
```

Comment:
This if is at the wrong place. You are mixing the condition on the inputs with the code to init the output.

This are 2 orthogonals thigs. If data is contiguous, then it goes to line 114, but out isn't created, so this segfault.